src/ILLink.Shared/XmlAttributeParser.cs

```diff
@@ -0,0 +1,347 @@
+			public string Fullname;
+			public AssemblyNode (XPathNavigator nav) : base (nav)
+			{
+				Fullname = GetFullName (nav);
+				Types = new List<TypeNode> ();
+				foreach (XPathNavigator typeNav in nav.SelectChildren (TypeElementName, "")) {
+					Types.Add (new TypeNode (typeNav));
+				}
+			}
+		}
+
+		public interface IRootNode
+		{
+		}
+
+		public abstract record NodeBase
+		{
+			protected const string FullNameAttributeName = "fullname";
+			protected const string LinkerElementName = "linker";
+			protected const string TypeElementName = "type";
+			protected const string SignatureAttributeName = "signature";
+			protected const string NameAttributeName = "name";
+			protected const string FieldElementName = "field";
+			protected const string MethodElementName = "method";
+			protected const string EventElementName = "event";
+			protected const string PropertyElementName = "property";
+			protected const string AttributeElementName = "attribute";
+			protected const string AssemblyElementName = "assembly";
+			protected const string ArgumentElementName = "argument";
+			protected const string ParameterElementName = "parameter";
+			protected const string ReturnElementName = "return";
+			protected const string AllAssembliesFullName = "*";
+			protected const string XmlNamespace = "";
+
+			public static List<IRootNode> ProcessRootNodes (XPathNavigator nav)
+			{
+				if (!nav.MoveToChild(LinkerElementName, XmlNamespace)) {
+					throw new ArgumentException ("XML does not have <linker> base tag");
+				}
+				var roots = new List<IRootNode> ();
+				foreach (XPathNavigator typeNav in nav.SelectChildren (TypeElementName, "")) {
+					roots.Add (new TypeNode (typeNav));
+				}
+				foreach (XPathNavigator assemblyNav in nav.SelectChildren (AssemblyElementName, "")) {
+					roots.Add (new TypeNode (assemblyNav));
+				}
+				return roots;
+			}
+			protected static string GetFullName (XPathNavigator nav)
+			{
+				return GetAttribute (nav, FullNameAttributeName);
+			}
+
+			protected static string GetName (XPathNavigator nav)
+			{
+				var name = GetAttribute (nav, NameAttributeName);
+				return name;
+			}
+
+			protected static string GetSignature (XPathNavigator nav)
+			{
+				return GetAttribute (nav, SignatureAttributeName);
+			}
+
+			protected static string GetAttribute (XPathNavigator nav, string attribute)
+			{
+				return nav.GetAttribute (attribute, XmlNamespace);
+			}
+
+		}
+
+		public abstract record AttributeTargetNode : NodeBase
+		{
+			public List<AttributeNode> Attributes;
+
+			public AttributeTargetNode (XPathNavigator nav)
+			{
+				var attributes = new List<AttributeNode> ();
+				foreach (XPathNavigator attributeNav in nav.SelectChildren (AttributeElementName, "")) {
+					var attr = new AttributeNode (attributeNav);
+					if (attr == null)
+						continue;
+					attributes.Add (attr);
+				}
+				Attributes = attributes;
+			}
+		}
+
+		public abstract record FeatureGatedNode : AttributeTargetNode
```

Comment:
I think we should change the name here. Main reason is that it's "new" - we don't use "feature gate" anywhere else. I understand that the meaning is slightly different to the widely used "feature switch", but maybe it would be better to keep the same terminology.

Reply:
Do you mean something like `FeatureSwitch` and `FeatureSwitchedNode`?

Reply:
`FeatureSwitchNode` could work. Or maybe a descriptive `NodeWithFeatureSwitch` ...

Technically in the linker we don't call them "feature switches", just "features", but that is a very generic name I think. so maybe "feature switches" in good enough.